src/main/java/com/amazon/opendistroforelasticsearch/security/filter/OpenDistroSecurityFilter.java

```diff
@@ -283,7 +293,30 @@ public int order() {
                 if(!dlsFlsValve.invoke(request, listener, pres.getAllowedFlsFields(), pres.getMaskedFields(), pres.getQueries())) {
                     return;
                 }
-                chain.proceed(task, action, request, listener);
+                final CreateIndexRequest createIndexRequest = pres.getRequest();
```

Comment:
Style: This is somewhat a breach of separation of concerns. This might lead to confusion (as illustrated by the existing comments) and hard to understand logic. One should try to concentrate the logic in PrivilegesInterceptorImpl. I can think of two alternatives:

- Provide a generic callback instead of the CreateIndexRequest
- Dynamically extend the ActionFilterChain. This would however probably require extension of a couple of method interfaces.

Reply:
In the current implementation `OpenDistroSecurityFilter` is responsible for executing the `chain` and pre-processing (submitting)`CreateIndexRequest` while `PrivilegesInterceptorImpl` is responsible for creating `CreateIndexRequest` instance if one is necessary.

I don't think that introducing callback will add more clarity, unless I miss something in your proposal. The same for `ActionFilterChain`, it is possible to return new instance of `ActionFilterChain` instead of `CreateIndexRequest` that will either call `client.admin().indices().create()` or `chain.proceed()` directly. This will require passing `chain` all the way to `PrivilegesInterceptorImpl` and propagating new `ActionFilterChain` back to `OpenDistroSecurityFilter`.

Reply:
So far, `OpenDistroSecurityFilter` was a very generic module and not concerned with multi-tenancy at all. All multi-tenancy logic was concentrated in `PrivilegesInterceptorImpl`. This separation would be no longer valid any more.

A callback would bring more clarity by still being able to keep the logic in `PrivilegesInterceptorImpl`. 

As I wrote, this is just a style issue. I can understand that callbacks would also complicate the code. When keeping the code in `OpenDistroSecurityFilter`, however, I'd recommend to clearly separate this code from the generic request handling and mark this as multi-tenancy handling.

Reply:
All logic related to multi-tenancy and whether or not it is necessary to create multi-tenancy index and details of such request is encapsulated in the `PrivilegesInterceptorImpl`. `OpenDistroSecurityFilter` is only responsible for submitting `CreateIndexRequest` and has no knowledge about multi-tenancy.